httpclient/client.go

```diff
@@ -496,40 +538,68 @@ func extractAndMergeChunks(chunksPaths []string, flags ConcurrentDownloadFlags,
 	return true, nil
 }
 
-func (jc *HttpClient) downloadFileRange(flags ConcurrentDownloadFlags, start, end int64, currentSplit int, logMsgPrefix string, httpClientsDetails httputils.HttpClientDetails, retries int) (string, error) {
+func (jc *HttpClient) downloadFileRangeWithRetries(flags ConcurrentDownloadFlags, start, end int64, currentSplit int, logMsgPrefix string,
+	httpClientsDetails httputils.HttpClientDetails, retries int) (fileName string, err error) {
+	var resp *http.Response
+	retryExecutor := retry.RetryExecutor{
+		MaxRetries:      retries,
+		RetriesInterval: 0,
+		RetryMessage:    fmt.Sprintf("Failure occurred while downloading part %d of %s", currentSplit, flags.DownloadPath),
+		LogMsgPrefix:    fmt.Sprintf("%s[%s]: ", logMsgPrefix, strconv.Itoa(currentSplit)),
+		ExecutionHandler: func() (bool, error) {
+			fileName, resp, err = jc.downloadFileRange(flags, start, end, currentSplit, logMsgPrefix, httpClientsDetails)
+			if err != nil {
+				return true, err
+			}
+
+			// If response-code < 500, should not retry
+			if resp != nil && resp.StatusCode < 500 {
+				return false, nil
+			}
+
+			// Perform retry
+			return true, nil
+		},
+	}
+
+	err = retryExecutor.Execute()
+	return
+}
+
+func (jc *HttpClient) downloadFileRange(flags ConcurrentDownloadFlags, start, end int64, currentSplit int, logMsgPrefix string, httpClientsDetails httputils.HttpClientDetails) (string, *http.Response, error) {
```

Comment:
It might be better to use named returned values in this method. This should make it easier to understand what the first returned value is.

Reply:
Changed to named-returned values.